### PR TITLE
Trigger service: initialize database command

### PR DESCRIPTION
**Pull request by rohanjr**
Minimal database initialization with schemas for `running_triggers` and `dalfs` tables. The user passes in the database URL, username and password in a config string argument (approach and code adapted from the JSON API).

In future the idea is to also create a "service" role with permissions to read and write to the new tables. Then the user can pass in the service role to connect to the database when running the service for real.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: triggers/service/BUILD.bazel [hunk 1]
```diff
@@ -47,6 +51,13 @@ da_scala_library(
         "@maven//:io_spray_spray_json_2_12",
         "@maven//:org_scalaz_scalaz_core_2_12",
         "@maven//:org_slf4j_slf4j_api",
+        "@maven//:org_tpolecat_doobie_core_2_12",
+        "@maven//:org_tpolecat_doobie_free_2_12",
+#        "@maven//:org_tpolecat_doobie_postgres_2_12",
```
Comment (shayne-fletcher) [inline triggers/service/BUILD.bazel]: Remove?

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerDao.scala [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+          trigger_name text not null
+        )
+      """
+    val createDarTable: Fragment = sql"""
+        create table dar_packages(
+          package_id text primary key,
+          package text not null
+        )
+      """
+//    val insertTrigger: Fragment =
```
Comment (shayne-fletcher) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerDao.scala]: Do you want to keep these comments for now?

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerDao.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        create table running_triggers(
+          trigger_id uuid primary key,
+          party_token text not null,
+          trigger_name text not null
+        )
+      """
+    val createDarTable: Fragment = sql"""
+        create table dar_packages(
+          package_id text primary key,
+          package text not null
```
Comment (cocreature) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerDao.scala:53]: Unless I am missing something here, this should be binary data not a unicode string, i.e., `bytea` in postgres I believe.

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala [hunk 4]
```diff
@@ -18,8 +19,55 @@ case class ServiceConfig(
...
+    jdbcConfig: Option[JdbcConfig],
 )
 
+final case class JdbcConfig(
+    url: String,
+    user: String,
+    password: String,
+)
+
+object JdbcConfig {
```
Comment (cocreature) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:32]: I assume you copied this from some other place in our codebase? Can we share this?

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala [hunk 5]
```diff
@@ -64,7 +112,19 @@ object ServiceConfig {
         c.copy(commandTtl = Duration.ofSeconds(t))
       }
       .text("TTL in seconds used for commands emitted by the trigger. Defaults to 30s.")
+
+    opt[Map[String, String]]("jdbc")
```
Comment (cocreature) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:116]: We now have `--jdbc` for the trigger service, `--query-store-jdbc-config` for the JSON API and `--sql-backend-jdbcurl` for sandbox the latter even has a different format :crying_cat_face:

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala [hunk 6]
```diff
@@ -64,7 +112,19 @@ object ServiceConfig {
...
+      .action((x, c) =>
+        c.copy(jdbcConfig = Some(JdbcConfig.create(x).fold(e => sys.error(e), identity))))
+      .optional()
+      .valueName(JdbcConfig.usage)
+      .text("JDBC configuration parameters. If omitted the service runs without a database.")
+
+    cmd("init-db")
```
Comment (cocreature) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:123]: This is a bit confusing. You still require all the parameters that are required if you do not use the command due to the way `scopt` works so you have to pass in the ledger-host even though it is not used.

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerDao.scala [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    val cs: ContextShift[IO] = IO.contextShift(ec)
+    new TriggerDao(Connection.connect(jdbcDriver, jdbcUrl, username, password)(cs))
+  }
+
+  def initialize(implicit log: LogHandler): ConnectionIO[Unit] = {
+    val createTriggerTable: Fragment = sql"""
+        create table running_triggers(
+          trigger_id uuid primary key,
+          party_token text not null,
+          trigger_name text not null
```
Comment (cocreature) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerDao.scala:47]: presumably this is not just the name but `package_id:module_name:trigger_name`. If you store them separately you save yourself from having to parse it everytime you read it from the DB. Either way, we should document what we store here.

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerDao.scala [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        )
+      """
+    val createDarTable: Fragment = sql"""
+        create table dar_packages(
+          package_id text primary key,
```
Comment (cocreature) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerDao.scala:52]: Your table is called `dar_packages` but one DAR has multiple dalfs and therefore package ids. I think you mean `dalf_packages` or just `dalfs`.

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala (hunk 4 above)
Comment (rohanjr) [reply triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:32]: Copied from the JSON API but slightly reduced to omit the driver (which can only be one thing in the JSON API as well) and a flag they use to indicate initialization in place of my `init-db` command. I'm sure there should be reuse here but I wanted to figure out the right format/interface for this thing first (including addressing your next comment).

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala (hunk 5 above)
Comment (rohanjr) [reply triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:116]: Good point, I could use the sandbox format and argument name, by including the username and password in the url string. I'll think about tweaking this in a later PR if that's ok.

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala (hunk 6 above)
Comment (rohanjr) [reply triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:123]: Agreed. The problem I ran into is that the `TriggerDao` object requires an implicit execution context (for connecting to the database), and that is obtained from the actor system which is initialized with those config parameters.. Perhaps I could instead use the global execution context in the `init-db` case?

**Comment by rohanjr** [comment]
Thanks for the review @cocreature ! I added a unit test for db initialization and made the schema changes you suggested.

**Comment by shayne-fletcher** [comment]
Good progress Rohan.

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala (hunk 5 above)
Comment (cocreature) [reply triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:116]: Happy to do it separately, I’d just like to not have to remember 3 different ways of specifying a jdbc url in the long term :slightly_smiling_face:

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala (hunk 6 above)
Comment (cocreature) [reply triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:123]: Or a custom one. I never really understood which execution context you are supposed to use tbh :slightly_smiling_face:. Happy to keep it for now.

**Pull request merged**